xarray/tests/test_missing.py

```diff
@@ -0,0 +1,441 @@
+                        dims='time')
+
+
+@pytest.fixture
+def ds():
+    ds = xr.Dataset()
+    ds['var1'] = xr.DataArray([0, np.nan, 1, 2, np.nan, 3, 4, 5, np.nan, 6, 7],
+                              dims='time')
+    ds['var2'] = xr.DataArray([10, np.nan, 11, 12, np.nan, 13, 14, 15, np.nan,
+                               16, 17], dims='x')
+    return ds
+
+
+def make_interpolate_example_data(shape, frac_nan, seed=12345,
+                                  non_uniform=False):
+    rs = np.random.RandomState(seed)
+    vals = rs.normal(size=shape)
+    if frac_nan == 1:
+        vals[:] = np.nan
+    elif frac_nan == 0:
+        pass
+    else:
+        n_missing = int(vals.size * frac_nan)
+
+        ys = np.arange(shape[0])
+        xs = np.arange(shape[1])
+        if n_missing:
+            np.random.shuffle(ys)
+            ys = ys[:n_missing]
+
+            np.random.shuffle(xs)
+            xs = xs[:n_missing]
+
+            vals[ys, xs] = np.nan
+
+    if non_uniform:
+        # construct a datetime index that has irregular spacing
+        deltas = pd.TimedeltaIndex(unit='d', data=rs.normal(size=shape[0],
+                                                            scale=10))
+        coords = {'time': (pd.Timestamp('2000-01-01') + deltas).sort_values()}
+    else:
+        coords = {'time': pd.date_range('2000-01-01', freq='D',
+                  periods=shape[0])}
+    da = xr.DataArray(vals, dims=('time', 'x'), coords=coords)
+    df = da.to_pandas()
+
+    return da, df
+
+
+@requires_np112
+@requires_scipy
+def test_interpolate_pd_compat():
+    shapes = [(8, 8), (1, 20), (20, 1), (100, 100)]
+    frac_nans = [0, 0.5, 1]
+    methods = ['linear',  'nearest', 'zero', 'slinear', 'quadratic', 'cubic']
+
+    for (shape, frac_nan, method) in itertools.product(shapes, frac_nans,
+                                                       methods):
+
+        da, df = make_interpolate_example_data(shape, frac_nan)
+
+        for dim in ['time', 'x']:
+            actual = da.interpolate_na(method=method, dim=dim)
+            expected = df.interpolate(method=method, axis=da.get_axis_num(dim))
+            np.testing.assert_allclose(actual.values, expected.values)
+
+
+@requires_np112
+@requires_scipy
+def test_scipy_methods_function():
+    for method in ['barycentric', 'krog', 'pchip', 'spline', 'akima']:
+        kwargs = {}
+        # Note: Pandas does some wacky things with these methods and the full
+        # integration tests wont work.
+        da, _ = make_interpolate_example_data((25, 25), 0.4, non_uniform=True)
+        actual = da.interpolate_na(method=method, dim='time', **kwargs)
+        assert (da.count('time') <= actual.count('time')).all()
+
+
+@requires_np112
+@requires_scipy
+def test_interpolate_pd_compat_non_uniform_index():
+    shapes = [(8, 8), (1, 20), (20, 1), (100, 100)]
+    frac_nans = [0, 0.5, 1]
+    methods = ['time', 'index', 'values']
```

Comment:
obv don't change but for future tests: pytest handles these parameterizations v well

Reply:
@maxim-lian - I was using the pytest parameterizations here but they weren't playing nice with the `@requires_foo` decorators.

Reply:
I can have a look. Do you remember what the setup was that didn't work? No stress if not

Reply:
I think skipped tests are returned as functions without arguments. The parameterizations then were passed to functions that didn't take arguments and that obviously didn't work.

So: 
```Python
@parameterize('foo', [1, 2])
@skip
def bar(foo):
    pass
```

becomes 
```Python
@parameterize('foo', [1, 2])
def bar():
    pass
```
when skipped.

Reply:
I think we should use `pytest.importorskip` - while @jreback is right that it [https://github.com/pandas-dev/pandas/issues/18190#issuecomment-353460062](doesn't allow as much control) as writing a function ourselves, and @WillAyd gives some [more benefits](https://github.com/pandas-dev/pandas/issues/18190#issuecomment-353460062) , we're not using anything that the standard pytest function can't do, in any of our functions.

If people agree, I could add a note to this function: https://github.com/pydata/xarray/pull/1640/files/48505a51e69d6b3663783db414962f3011e94aa1#diff-e84d0339b6da4d18c02f6b2d54fbd5bdR47

Reply:
I recall we had some difficulty with decorators based on pytest and subclass methods, but that doesn't apply here.